lib/memoist.rb

```diff
@@ -63,28 +63,49 @@ def unmemoize_all
       flush_cache
     end
 
+    def init_lazy_memoized_structs(names)
+      structs = self.class.memoized_methods
+      return structs if names.empty?
+
+      structs.select { |s| names.include?(s.memoized_method) }
+    end
+
     def prime_cache(*method_names)
-      method_names = self.class.memoized_methods if method_names.empty?
-      method_names.each do |method_name|
-        if method(Memoist.unmemoized_method_for(method_name)).arity == 0
-          __send__(method_name)
+      memoized_structs = init_lazy_memoized_structs(method_names)
+      memoized_structs.each do |memoized_struct|
+        if method(memoized_struct.unmemoized_method).arity == 0
+          __send__(memoized_struct.memoized_method)
         else
-          ivar = Memoist.memoized_ivar_for(method_name)
-          instance_variable_set(ivar, {})
+          instance_variable_set(memoized_struct.ivar, {})
         end
       end
     end
 
     def flush_cache(*method_names)
-      method_names = self.class.memoized_methods if method_names.empty?
-
-      method_names.each do |method_name|
-        ivar = Memoist.memoized_ivar_for(method_name)
-        remove_instance_variable(ivar) if instance_variable_defined?(ivar)
+      memoized_structs = init_lazy_memoized_structs(method_names)
+      memoized_structs.each do |memoized_struct|
+        remove_instance_variable(memoized_struct.ivar) if instance_variable_defined?(memoized_struct.ivar)
       end
     end
   end
 
+  MemoizedMethodStruct = Struct.new(:memoized_method, :ivar, :unmemoized_method) do
```

Comment:
Naming, it's hard... It's better than my first one: MemoizedMethodThing @matthewd